File: website/utils/expenses.py
from sqlalchemy import create_engine, MetaData, Table, text
from sqlalchemy.orm import sessionmaker
import re

ETJ_TAX_RATE = {
    'DURHAM CITY': 1.3099,  # $1.3099 per $100 of assessed value
    'DURHAM COUNTY': 0.7522,  # $0.7522 per $100 of assessed value
    'DURHAM COUNTY,DURHAM CITY': 1.3099,  # Same as Durham City
    'CHAPEL-HILL': 0.82,  # Approximate, adjust if needed
    'RALEIGH': 0.98,  # Approximate, adjust if needed
    'DURHAM COUNTY,RALEIGH': 0.98,  # Follows Raleigh's rate
    'DURHAM COUNTY,MORRISVILLE': 0.75,  # Lower estimate
    'MORRISVILLE': 0.75,  # Estimated for Morrisville
    'CHAPEL-HILL,DURHAM CITY': 1.3099,  # Same as Durham City
    'CARY': 1.15,  # Approximate rate
    'CHAPEL-HILL,DURHAM COUNTY': 0.82  # Uses Chapel-Hill approximation
}

MANAGEMENT_RESERVE_PCT = 8/100
MAINTENENCE_RESERVE_PCT = 4/100

ELECTRIC_PER_KWH = 13.72
KWH_USAGE_PER_SQFT = 12 # can be up to around 15
ELECTRIC_PER_SQFT = ELECTRIC_PER_KWH*KWH_USAGE_PER_SQFT

WATER_TIERS = {
    2: 2.45, # TOP LIMIT USAGE CCF: $ PER CCF
    5: 3.69,
    8: 4.05,
    15: 5.28,
    "else": 7.4
}
SEWER_COST_PER_CCF = 5.26 # $ PER CCF WATER USAGE
WATER_USAGE_PER_OCCUPANT = (3,4)

GARBAGE_COST_PER_UNIT = 24 # PER UNIT

INSURANCE_COST_PCT = 0.5/100 # PERCENT OF TOTAL PROPERTY VALUE

NATURAL_GAS_PER_SQFT_PER_HR = 0.00049836
ELECTRIC_PER_SQFT_PER_HR = 0.0015463

LAWN_CARE_PER_ACRE = (100,1600) # RANGE FROM MOWING TWICE A WEEK TO LUXURY FULL SERVICE

LEGAL_AND_PROFESSIONAL_PCT = 1.5/100 # FOR MULTI-FAMILY HOMES 

def get_tables(db_path):
    # Create the engine
    engine = create_engine(f'sqlite:///{db_path}')

    # Reflect the existing database into a new MetaData object
    metadata = MetaData()
    metadata.reflect(bind=engine)

    # Print all table names
    print(metadata.tables.keys())
    

def get_db_schema(db_path):
    
    # Create an engine that connects to the SQLite database
    engine = create_engine(f'sqlite:///{db_path}')

    # Create a session
    Session = sessionmaker(bind=engine)
    session = Session()

    # Reflect the tables from the database
    metadata = MetaData()
    metadata.reflect(bind=engine)

    # You can now access the tables and work with them
    for table_name, table in metadata.tables.items():
        print(f'Table: {table_name}')
        for column in table.columns:
            print(f'Column: {column.name}, Type: {column.type}')

    # Example: querying a table (replace 'your_table' with an actual table name)
    your_table = Table('your_table_name', metadata, autoload_with=engine)
    results = session.query(your_table).all()

    # for row in results:
    #     print(row)
    
def query_db(db_path, query):
    engine = create_engine(f'sqlite:///{db_path}')
    with engine.connect() as connection:
        result = connection.execute(text(query))
        rows = result.fetchall()
    for row in rows:
        print(row)
    print(f"\n There are {len(rows)} rows")

def estimate_water_cost(num_occupants, water_usage_per_occupant):
    total_usage = water_usage_per_occupant * num_occupants
    teirs = [val if type(val) == int else float("inf") for val in WATER_TIERS.keys()]
    teirs.sort()
    uncalculated_usage = total_usage
    # print(total_usage)
    cost = total_usage * SEWER_COST_PER_CCF # calcualate sewer usage
    for i in range(len(teirs)):
        if i < len(teirs) - 1:
            ccf_to_pay = min(teirs[i]-teirs[i-1], uncalculated_usage) if i > 0 else min(teirs[i]-0, uncalculated_usage)
            cost += ccf_to_pay * WATER_TIERS[teirs[i]]
            uncalculated_usage -= ccf_to_pay
        else:
            if uncalculated_usage > 0:
                cost += uncalculated_usage*WATER_TIERS['else']
    return cost

def estimate_garbage_cost(num_units, include):
    return GARBAGE_COST_PER_UNIT * num_units if include else False

def estimate_units(zoning_code, lot_size_acres=1):
    """
    Estimates the number of residential units based on the zoning code and lot size.
    
    Args:
        zoning_code (str): The zoning code (e.g., 'RS-10', 'PDR 6.050', 'RU-5').
        lot_size_acres (float): The total size of the land in acres (default is 1 acre).
    
    Returns:
        int: Estimated number of residential units for the given zoning code.
    """
    # Handle PDR zones (e.g., 'PDR 6.050')
    pdr_match = re.match(r'PDR (\d+\.\d+)', zoning_code)
    if pdr_match:
        units_per_acre = float(pdr_match.group(1))
        return max(int(units_per_acre * lot_size_acres), 1)

    # Handle RS zones (e.g., 'RS-10', 'RS-8')
    rs_match = re.match(r'RS-(\d+)', zoning_code)
    if rs_match:
        min_lot_size_sqft = int(rs_match.group(1)) * 1_000  # Convert to square feet
        units = (lot_size_acres * 43_560) // min_lot_size_sqft  # 43,560 sq ft in an acre
        return max(1, int(units))  # Ensure at least one unit is returned

    # Handle RU zones (e.g., 'RU-5', 'RU-5(2)')
    ru_match = re.match(r'RU-5(?:\((\d)\))?', zoning_code)
    if ru_match:
        multiplier = int(ru_match.group(1)) if ru_match.group(1) else 1
        units = (lot_size_acres * 43_560) // (5_000 / multiplier)
        return max(1, int(units))

    # Handle special or mixed-use zones
    if 'MU' in zoning_code or 'CC' in zoning_code:
        return max(int(10 * lot_size_acres), 1)  # Assume 10 units per acre for mixed-use areas

    # Default for unhandled zoning codes
    return 1

def estimate_insurance_cost(property_value):
    return INSURANCE_COST_PCT*property_value

def estimate_heating_cost(isGas, sqft):
    avg_heat_running_hrs = 5
    days_in_month = 30
    return NATURAL_GAS_PER_SQFT_PER_HR*avg_heat_running_hrs*days_in_month*sqft if isGas else ELECTRIC_PER_SQFT_PER_HR*avg_heat_running_hrs*days_in_month*sqft
            
def estimate_legal_cost(annual_rental_income, num_units):
    if num_units > 1:
        return annual_rental_income*.015
    else:
        return 0

def management_reserve_cost(property_value):
    return property_value*MANAGEMENT_RESERVE_PCT

def maintenence_reserve_cost(property_value):
    return property_value*MAINTENENCE_RESERVE_PCT
    
def calculate_property_tax(etj, property_value):
    if etj is None or etj not in ETJ_TAX_RATE:
        raise ValueError("Invalid or unknown ETJ code provided.")

    tax_rate = ETJ_TAX_RATE[etj]

    property_tax = (tax_rate / 100) * property_value

    return round(property_tax, 2)

def calculate_lawn_snow_cost(acres, lawn_care_per_acre):
    return acres*lawn_care_per_acre
    
def estimate_expenses(annual_rental_income, num_units, isGas, sqft, property_value, num_beds, etj, acres):
    sqft = 625 * num_beds
    costs = [
        estimate_legal_cost(annual_rental_income, num_units),
        estimate_heating_cost(isGas, sqft),
        estimate_insurance_cost(property_value),
        estimate_garbage_cost(num_units, include=True),
        estimate_water_cost(num_beds, water_usage_per_occupant=sum(WATER_USAGE_PER_OCCUPANT)/len(WATER_USAGE_PER_OCCUPANT)),
        management_reserve_cost(property_value),
        maintenence_reserve_cost(property_value),
        calculate_property_tax(etj, property_value),
        calculate_lawn_snow_cost(acres, lawn_care_per_acre=sum(LAWN_CARE_PER_ACRE)/len(LAWN_CARE_PER_ACRE))
    ]
    print(costs)
    return sum(costs)

File: website/static/expenses.js
// estimateExpenses.js

const ETJ_TAX_RATE = {
    'DURHAM CITY': 1.3099,
    'DURHAM COUNTY': 0.7522,
    'DURHAM COUNTY,DURHAM CITY': 1.3099,
    'CHAPEL-HILL': 0.82,
    'RALEIGH': 0.98,
    'DURHAM COUNTY,RALEIGH': 0.98,
    'DURHAM COUNTY,MORRISVILLE': 0.75,
    'MORRISVILLE': 0.75,
    'CHAPEL-HILL,DURHAM CITY': 1.3099,
    'CARY': 1.15,
    'CHAPEL-HILL,DURHAM COUNTY': 0.82
};

const MANAGEMENT_RESERVE_PCT = 0.08;
const MAINTENENCE_RESERVE_PCT = 0.04;

const ELECTRIC_PER_KWH = 13.72;
const KWH_USAGE_PER_SQFT = 12;
const ELECTRIC_PER_SQFT = ELECTRIC_PER_KWH * KWH_USAGE_PER_SQFT;

const WATER_TIERS = { 2: 2.45, 5: 3.69, 8: 4.05, 15: 5.28, "else": 7.4 };
const SEWER_COST_PER_CCF = 5.26;
const WATER_USAGE_PER_OCCUPANT = [3, 4];

const GARBAGE_COST_PER_UNIT = 24;
const INSURANCE_COST_PCT = 0.005;
const NATURAL_GAS_PER_SQFT_PER_HR = 0.00049836;
const ELECTRIC_PER_SQFT_PER_HR = 0.0015463;
const LAWN_CARE_PER_ACRE = [100, 1600];
const LEGAL_AND_PROFESSIONAL_PCT = 0.015;

function estimateWaterCost(numOccupants) {
    const avgUsage = (WATER_USAGE_PER_OCCUPANT[0] + WATER_USAGE_PER_OCCUPANT[1]) / 2;
    let totalUsage = avgUsage * numOccupants;
    let cost = totalUsage * SEWER_COST_PER_CCF;
    let remainingUsage = totalUsage;

    Object.keys(WATER_TIERS).sort((a, b) => a - b).forEach(limit => {
        if (remainingUsage > 0) {
            const tierLimit = (typeof limit === 'number') ? parseFloat(limit) : Infinity;
            const usageInTier = Math.min(tierLimit, remainingUsage);
            cost += usageInTier * WATER_TIERS[limit];
            remainingUsage -= usageInTier;
        }
    });
    return cost;
}

function estimateGarbageCost(numUnits, include) {
    return include ? GARBAGE_COST_PER_UNIT * numUnits : 0;
}

function estimateUnits(zoningCode, lotSizeAcres = 1) {
    if (zoningCode.startsWith('PDR')) {
        const unitsPerAcre = parseFloat(zoningCode.split(' ')[1]);
        return Math.max(Math.floor(unitsPerAcre * lotSizeAcres), 1);
    }
    if (zoningCode.startsWith('RS')) {
        const minLotSizeSqft = parseInt(zoningCode.split('-')[1]) * 1000;
        return Math.max(Math.floor((lotSizeAcres * 43560) / minLotSizeSqft), 1);
    }
    if (zoningCode.startsWith('RU-5')) {
        const multiplier = parseInt(zoningCode.split('(')[1] || 1);
        return Math.max(Math.floor((lotSizeAcres * 43560) / (5000 / multiplier)), 1);
    }
    if (zoningCode.includes('MU') || zoningCode.includes('CC')) {
        return Math.max(Math.floor(10 * lotSizeAcres), 1);
    }
    return 1;
}

function estimateInsuranceCost(propertyValue) {
    return INSURANCE_COST_PCT * propertyValue;
}

function estimateHeatingCost(isGas, sqft) {
    const avgHours = 5;
    const daysInMonth = 30;
    return isGas ? NATURAL_GAS_PER_SQFT_PER_HR * avgHours * daysInMonth * sqft : ELECTRIC_PER_SQFT_PER_HR * avgHours * daysInMonth * sqft;
}

function estimateLegalCost(annualRentalIncome, numUnits) {
    return numUnits > 1 ? annualRentalIncome * LEGAL_AND_PROFESSIONAL_PCT : 0;
}

function managementReserveCost(propertyValue) {
    return propertyValue * MANAGEMENT_RESERVE_PCT;
}

function maintenenceReserveCost(propertyValue) {
    return propertyValue * MAINTENENCE_RESERVE_PCT;
}

function calculatePropertyTax(etj, propertyValue) {
    const taxRate = ETJ_TAX_RATE[etj] || 0;
    return (taxRate / 100) * propertyValue;
}

function calculateLawnSnowCost(acres) {
    const avgLawnCareCost = (LAWN_CARE_PER_ACRE[0] + LAWN_CARE_PER_ACRE[1]) / 2;
    return acres * avgLawnCareCost;
}

function estimateExpenses(annualRentalIncome, numUnits, isGas, sqft, propertyValue, numBeds, etj, acres, legal, utilities, maintenencePct, managementPct) {
    const costs = [
        legal ? legal: estimateLegalCost(annualRentalIncome, numUnits),
        utilities ? utilities: estimateHeatingCost(isGas, sqft),
        utilities ? 0: estimateInsuranceCost(propertyValue),
        utilities ? 0: estimateGarbageCost(numUnits, true),
        utilities ? 0: estimateWaterCost(numBeds),
        managementPct ? managementPct*annualRentalIncome : managementReserveCost(annualRentalIncome),
        maintenencePct ? maintenencePct*annualRentalIncome : maintenenceReserveCost(annualRentalIncome),
        calculatePropertyTax(etj, propertyValue),
        utilities ? 0: calculateLawnSnowCost(acres)
    ];
    console.log(estimateLegalCost(annualRentalIncome, numUnits))
    return costs.reduce((total, cost) => total + cost, 0);
}

// function gradeHouse(capRate, noi, propertyValue, monthlyRent, vacancyAllowance, numUnits, sqft, neighborhoodScore, expenses, zoningScore) {
    

//     // Define weights for each factor
//     const weights = {
//         capRate: 0.03,          // Reduce by factor of 10
//         noi: 0.02,              // Reduce by factor of 10
//         propertyValue: -0.01,   // Reduce by factor of 10
//         monthlyRent: 0.02,      // Reduce by factor of 10
//         vacancyAllowance: -0.005, // Reduce by factor of 10
//         numUnits: 0.01,         // Reduce by factor of 10
//         sqft: 0.005,            // Reduce by factor of 10
//         neighborhood: 0.01,     // Reduce by factor of 10
//         expenses: -0.02,        // Reduce by factor of 10
//         zoning: 0.005           // Reduce by factor of 10
//     };
    

//     // Calculate the score with weights applied
//     const score = (
//         weights.capRate * capRate +
//         weights.noi * noi +
//         weights.propertyValue * (1 / propertyValue) +  // Inverse for lower value preference
//         weights.monthlyRent * monthlyRent +
//         weights.vacancyAllowance * (1 - vacancyAllowance) +
//         weights.numUnits * numUnits +
//         weights.sqft * sqft +
//         // weights.neighborhood * neighborhoodScore +
//         weights.expenses * (1 / expenses)  // Inverse for lower expenses
//         // weights.zoning * zoningScore
//     );

//     // Define realistic min and max scores for normalization
//     const minScore = 0;  // Minimum expected scaled-down score
//     const maxScore = 10; // Maximum expected scaled-down score

//     // Normalize to 1–100
//     const normalizedScore = 1 + ((score - minScore) / (maxScore - minScore)) * 99;
//     // const finalScore = Math.min(100, Math.max(1, normalizedScore));
//     const k = -1/30
//     const x0 = 100
//     const finalScore = 100 / (1 + Math.exp(k*(score - x0)))
    

//     // Assign grade based on score ranges
//     let grade;
//     if (finalScore >= 90) {
//         grade = 'A';
//     } else if (finalScore >= 75) {
//         grade = 'B';
//     } else if (finalScore >= 60) {
//         grade = 'C';
//     } else if (finalScore >= 40) {
//         grade = 'D';
//     } else {
//         grade = 'F';
//     }

//     return { grade, score, finalScore };
// }

function gradeHouse(capRate, noi, propertyValue, monthlyRent, vacancyAllowance, numUnits, sqft, neighborhoodScore, expenses, zoningScore) {
    // Industry standard metrics
    const annualRent = monthlyRent * 12 * numUnits;
    const rentPerSqft = (monthlyRent * 12) / sqft;
    const expenseRatio = expenses / annualRent;
    const pricePerUnit = propertyValue / numUnits;
    const pricePerSqft = propertyValue / sqft;
    
    // Scoring criteria based on industry standards
    const metrics = {
        capRate: {
            weight: 0.25,
            score: () => {
                // Cap rate scoring (relaxed range 3-8%)
                if (capRate < 1) return 0;
                if (capRate > 6) return 100;
                return ((capRate - 1) / 5) * 100;
            }
        },
        expenseRatio: {
            weight: 0.15,
            score: () => {
                // Expense ratio scoring (relaxed range 35-60%)
                if (expenseRatio > 0.75) return 0;
                if (expenseRatio < 0.35) return 100;
                return ((0.75 - expenseRatio) / 0.40) * 100;
            }
        },
        rentPerSqft: {
            weight: 0.15,
            score: () => {
                // Rent per sqft scoring (lowered thresholds)
                const annualRentPerSqft = rentPerSqft;
                if (annualRentPerSqft < 8) return 0;
                if (annualRentPerSqft > 24) return 100;
                return ((annualRentPerSqft - 8) / 16) * 100;
            }
        },
        pricePerUnit: {
            weight: 0.15,
            score: () => {
                // Price per unit scoring (increased range)
                if (pricePerUnit > 350000) return 0;
                if (pricePerUnit < 75000) return 100;
                return ((350000 - pricePerUnit) / 275000) * 100;
            }
        },
        vacancyRate: {
            weight: 0.15,
            score: () => {
                // Vacancy allowance scoring (relaxed range 5-15%)
                if (vacancyAllowance > 15) return 0;
                if (vacancyAllowance < 5) return 100;
                return ((15 - vacancyAllowance) / 10) * 100;
            }
        },
        cashOnCash: {
            weight: 0.15,
            score: () => {
                // Cash on Cash Return (lowered thresholds)
                const downPayment = propertyValue * 0.25;
                const cashOnCash = (noi / downPayment) * 100;
                if (cashOnCash < 5) return 0;
                if (cashOnCash > 15) return 100;
                return ((cashOnCash - 5) / 10) * 100;
            }
        }
    };

    // Calculate weighted score
    let weightedTotal = 0;
    let weightSum = 0;

    for (const [key, metric] of Object.entries(metrics)) {
        const metricScore = metric.score();
        weightedTotal += metricScore * metric.weight;
        weightSum += metric.weight;
    }

    // Calculate final score (0-100 scale)
    const finalScore = weightedTotal / weightSum;

    // More generous grade ranges
    let grade;
    if (finalScore >= 85) {
        grade = finalScore >= 95 ? 'A+' : finalScore >= 90 ? 'A' : 'A-';
    } else if (finalScore >= 70) {
        grade = finalScore >= 80 ? 'B+' : finalScore >= 75 ? 'B' : 'B-';
    } else if (finalScore >= 55) {
        grade = finalScore >= 65 ? 'C+' : finalScore >= 60 ? 'C' : 'C-';
    } else if (finalScore >= 40) {
        grade = finalScore >= 50 ? 'D+' : finalScore >= 45 ? 'D' : 'D-';
    } else {
        grade = 'F';
    }

    // Return detailed scoring breakdown
    return {
        grade,
        score: weightedTotal,
        finalScore,
        metrics: Object.entries(metrics).reduce((acc, [key, metric]) => {
            acc[key] = {
                score: metric.score(),
                weight: metric.weight
            };
            return acc;
        }, {})
    };
}